m365-developer-proxy-plugins/Inspection/DevToolsPlugin.cs

```diff
@@ -0,0 +1,299 @@
+        pluginEvents.AfterRequestLog += AfterRequestLog;
+    }
+
+    private static int GetFreePort()
+    {
+        var listener = new TcpListener(IPAddress.Loopback, 0);
+        listener.Start();
+        var port = ((IPEndPoint)listener.LocalEndpoint).Port;
+        listener.Stop();
+        return port;
+    }
+
+    private string GetChromiumProcessName(DevToolsPluginConfiguration configuration)
+    {
+        switch (configuration.PreferredBrowser)
+        {
+            case PreferredBrowser.Chrome:
+                return RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+                    ? "chrome.exe"
+                    : "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome";
+            case PreferredBrowser.EdgeDev:
+                return RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+                    ? Environment.ExpandEnvironmentVariables(@"%ProgramFiles(x86)%\Microsoft\Edge Dev\Application\msedge.exe")
+                    : "/Applications/Microsoft Edge Dev.app/Contents/MacOS/Microsoft Edge Dev";
+            default:
+                return RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
+                    ? Environment.ExpandEnvironmentVariables(@"%ProgramFiles(x86)%\Microsoft\Edge\Application\msedge.exe")
+                    : "/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge";
+        }
+    }
+
+    private void InitInspector()
+    {
+        var port = GetFreePort();
+        webSocket = new WebSocketServer(port);
+        webSocket.MessageReceived += SocketMessageReceived;
+        webSocket.Start();
+
+        var processName = GetChromiumProcessName(_configuration);
+        var inspectionUrl = $"http://localhost:9222/devtools/inspector.html?ws=localhost:{port}";
+        var args = $"{inspectionUrl} --remote-debugging-port=9222";
+
+        _logger?.LogInfo($"DevTools available at {inspectionUrl}");
+
+        var process = new Process
+        {
+            StartInfo = new()
+            {
+                FileName = processName,
+                Arguments = args,
+                // suppress default output
+                RedirectStandardError = true,
+                RedirectStandardOutput = true,
+                UseShellExecute = false
+            }
+        };
+        process.Start();
+    }
+
+    private void SocketMessageReceived(string msg)
+    {
+        if (webSocket is null)
+        {
+            return;
+        }
+
+        try
+        {
+            var message = JsonSerializer.Deserialize<GetResponseBodyMessage>(msg);
+            if (message?.Method == "Network.getResponseBody")
+            {
+                var requestId = message.Params?.RequestId;
+                if (requestId is null || !responseBody.ContainsKey(requestId))
+                {
+                    return;
+                }
+
+                var result = new GetResponseBodyResult
+                {
+                    Id = (int)message.Id!,
```

Comment:
Should we be checking for the Id on the message in the early exit conditions on 120? or providing some form of logical default?

Reply:
We could do early exit but it's a part of the spec that a message contains an ID which is why we're assuming it here. If dev tools would send us a message without it, they'd be breaking spec and we wouldn't be able to answer it because messages are matched on ID.